### ISSUE TITLE
SINCOS

### ISSUE THREAD

**Issue by jcarpent** [stack-of-tasks/pinocchio#810]
I've just opened this new issue to discuss the point of how to handle the SINCOS operation in the most generic to handle all the Scalar types.

```c++
    template<typename Scalar>
    struct SINCOSAlgo
    {
      static void run(const Scalar & a, Scalar * sa, Scalar * ca) 
      {   
        (*sa) = std::sin(a); (*ca) = std::cos(a);
      }   
    };
```
It would be more appropriate to just do `sin(a)`, `cos(a)` and let [argument-dependent lookup](https://en.cppreference.com/w/cpp/language/adl) do the job. For `casadi::SX` argument type, the calls would resolve to `casadi::sin()` and `casadi::cos()`. Notice that if we rely on ADL, the following code fragment becomes unnecessary:
```c++
    #ifdef PINOCCHIO_WITH_CPPAD_SUPPORT
  
    /// Implementation for CppAD scalar types.
    template<typename Scalar>
    struct SINCOSAlgo< CppAD::AD<Scalar> >
    {
      static void run(const CppAD::AD<Scalar> & a, CppAD::AD<Scalar> * sa, CppAD::AD<Scalar> * ca)
      {
        (*sa) = CppAD::sin(a); (*ca) = CppAD::cos(a);
      }
    };
  
    #endif
```

_Originally posted by @mkatliar in https://github.com/stack-of-tasks/pinocchio/issues/809#issuecomment-499509614_

**Comment by jcarpent** [comment]
To really take care of both classic scalar types (double, float, ...) as well as overloaded types (CppAD::AD or casadi::SX).
Originally, we were using the sincos instruction to compute `sin` and `cos` at the same time, which is much more efficient than computing the two separated.
Of course, some recent compilers are able to guess it. 
So, maybe it would be worth to just allows ADL to work for each types correctly, by checking if the types are standard or not.

**Comment by gabrielebndn** [comment]
There is a function is_fundamental in boost. Maybe the solution is to add a second template argument, which is boolean and computed through is_fundamental

**Comment by mkatliar** [comment]
@gabrielebndn why would we need `is_fundamental()`? `SINCOSAlgo` is already specialized for `float`, `double` and `long double`. For other types (like `casadi::SX` of `CppAD::AD<Scalar>`) we can assume that the `sin()` and `cos()` are defined in corresponding namespaces.

**Comment by gabrielebndn** [comment]
@mkatliar, I wanted to be generic with respect to any primitive type, but in effect float, double and long double cover all possible types of real numbers in C++ and it would not make sense at all to use integers, so that's probably useless, you are right

**Comment by jcarpent** [comment]
Solved by #812.
